execution/union_all.go

```diff
@@ -1,87 +1,120 @@
 }
 
 func (node *UnionAll) Get(ctx context.Context, variables octosql.Variables, streamID *StreamID) (RecordStream, error) {
-	tx := storage.GetStateTransactionFromContext(ctx)
-	prefixedTx := tx.WithPrefix(streamID.AsPrefix())
+	prefixedTx := storage.GetStateTransactionFromContext(ctx).WithPrefix(streamID.AsPrefix())
 
-	firstSourceStreamID, err := GetSourceStreamID(prefixedTx, octosql.MakeInt(1))
-	if err != nil {
-		return nil, errors.Wrap(err, "couldn't get first source stream ID")
-	}
-	secondSourceStreamID, err := GetSourceStreamID(prefixedTx, octosql.MakeInt(2))
-	if err != nil {
-		return nil, errors.Wrap(err, "couldn't get first source stream ID")
-	}
-
-	firstRecordStream, err := node.first.Get(ctx, variables, firstSourceStreamID)
-	if err != nil {
-		return nil, errors.Wrap(err, "couldn't get first record stream")
-	}
-	secondRecordStream, err := node.second.Get(ctx, variables, secondSourceStreamID)
-	if err != nil {
-		return nil, errors.Wrap(err, "couldn't get second record stream")
+	sourceRecordStreams := make([]RecordStream, len(node.sources))
+	for i := range node.sources {
+		sourceStreamID, err := GetSourceStreamID(prefixedTx, octosql.MakeInt(i))
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get source stream ID for source with index %d", i)
+		}
+		recordStream, err := node.sources[i].Get(ctx, variables, sourceStreamID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get source record stream with index %d", i)
+		}
+		sourceRecordStreams[i] = recordStream
 	}
 
 	return &UnifiedStream{
-		first:  firstRecordStream,
-		second: secondRecordStream,
+		sources:  sourceRecordStreams,
+		streamID: streamID,
 	}, nil
 }
 
 type UnifiedStream struct {
-	first, second RecordStream
+	sources  []RecordStream
+	streamID *StreamID
 }
 
 func (node *UnifiedStream) Close() error {
-	err := node.first.Close()
-	if err != nil {
-		return errors.Wrap(err, "Couldn't close first underlying stream")
-	}
-
-	err = node.second.Close()
-	if err != nil {
-		return errors.Wrap(err, "Couldn't close second underlying stream")
+	for i := range node.sources {
+		err := node.sources[i].Close()
+		if err != nil {
+			return errors.Wrapf(err, "couldn't close source stream with index %d", i)
+		}
 	}
 
 	return nil
 }
 
+var endsOfStreamsPrefix = []byte("$end_of_streams$")
+
 func (node *UnifiedStream) Next(ctx context.Context) (*Record, error) {
-	for {
-		firstRecord, err := node.first.Next(ctx)
-		if err != nil {
-			if err == ErrEndOfStream {
-				break
+	tx := storage.GetStateTransactionFromContext(ctx).WithPrefix(node.streamID.AsPrefix())
+	endOfStreamsMap := storage.NewMap(tx.WithPrefix(endsOfStreamsPrefix))
+
+	changeSubscriptions := make([]*storage.Subscription, len(node.sources))
+	for i := range node.sources { // TODO: Think about randomizing order.
+		key := octosql.MakeString(fmt.Sprint(i))
+		var endOfStream octosql.Value
+		err := endOfStreamsMap.Get(&key, &endOfStream)
+		if err == storage.ErrNotFound {
```

Comment:
Once again about the comments: before the whole `err := endOfStreamsMap...`
you could just add a comment "check if given stream source has already ended".
It would improve readability a lot